include/swift/Subsystems.h

```diff
@@ -286,29 +286,31 @@ namespace swift {
                                    StringRef OutputPath);
 
   /// Turn the given LLVM module into native code and return true on error.
-  bool performLLVM(const IRGenOptions &Opts, ASTContext &Ctx, llvm::Module *Module,
-                   StringRef OutputFilename,
-                   UnifiedStatsReporter *Stats=nullptr);
+  bool performLLVM(const IRGenOptions &Opts,
+                   ASTContext &Ctx,
+                   llvm::Module *Module,
+                   StringRef OutputFilename);
 
   /// Run the LLVM passes. In multi-threaded compilation this will be done for
   /// multiple LLVM modules in parallel.
-  /// \param Diags may be null if LLVM code gen diagnostics are not required.
-  /// \param DiagMutex may also be null if a mutex around \p Diags is not
-  ///                  required.
+  /// \param Diags The Diagnostic Engine.
+  /// \param DiagMutex in contexts that require parallel codegen, a mutex that the
+  ///                  diagnostic engine uses to synchronize emission.
   /// \param HashGlobal used with incremental LLVMCodeGen to know if a module
   ///                   was already compiled, may be null if not desired.
   /// \param Module LLVM module to code gen, required.
   /// \param TargetMachine target of code gen, required.
   /// \param effectiveLanguageVersion version of the language, effectively.
   /// \param OutputFilename Filename for output.
-  bool performLLVM(const IRGenOptions &Opts, DiagnosticEngine *Diags,
+  bool performLLVM(const IRGenOptions &Opts,
+                   DiagnosticEngine &Diags,
                    llvm::sys::Mutex *DiagMutex,
                    llvm::GlobalVariable *HashGlobal,
                    llvm::Module *Module,
                    llvm::TargetMachine *TargetMachine,
                    const version::Version &effectiveLanguageVersion,
                    StringRef OutputFilename,
-                   UnifiedStatsReporter *Stats=nullptr);
+                   UnifiedStatsReporter *Stats);
```

Comment:
If this can ever be null, I suggest using `NullablePtr<UnifiedStatsReporter>`

Reply:
We spoke in a different forum about this, but to have it out in public: It's probably a good idea to have the compiler use `UnifiedStatsReporter &` as a currency type for all of its interfaces. We can use the pimpl pattern or something to abstract away the conditional initialization. Or I oughta just go ahead and push the `NullablePtr` stuff through which would change like 200 callsites...